Add specific types for Fate locks and Zoo locks

Manno15 opened the pull request:
A possible solution for #1966.  

Giving lock paths a more specific type over just `String` will reduce confusion on which type of lock is being used and which appropriate function calls to make.

This is just one possible implementation of this. I wavered on including a specific path function or not. Any ideas on improvements are welcomed.

Manno15 commented:
The extra commits are from merging these changes from an alternate branch I had.

EdColeman commented on core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java:
```diff
@@ -60,6 +60,19 @@ public String toString() {
+      return this.path;
+    }
+  }
+
```
Just a general observation - would it improve readability if in addition to toString() that there was a getter (getPath())? 

Also, did you notice places where the node name was necessary?  If so, then would things be cleaner if there was a getNode() or something that did the parsing?  Might save a few lines here and there - but would gather that code into a single place.

Manno15 replied:
> Just a general observation - would it improve readability if in addition to toString() that there was a getter (getPath())?

I am not sure if it would improve readability. It could increase robustness but I do not think it benefits that much. 

> Also, did you notice places where the node name was necessary? If so, then would things be cleaner if there was a getNode() or something that did the parsing? Might save a few lines here and there - but would gather that code into a single place.

Most replacements were just for the string with this new path object (and then having toString when necessary). I did not notice anything about node name while making these changes.

ctubbsii replied:
@EdColeman These shouldn't be thought of as an object that *has* a path... rather, they *are* path types. So, adding `getPath()` to these would make them less concise and somewhat confusing (a path type with a getter for a path type is confusing to me, anyway). I think having a proper `toString()` here is sufficient.

EdColeman commented on core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java:
```diff
@@ -106,11 +119,11 @@ public ZooLock(AccumuloConfiguration conf, String path, UUID uuid) {
   }
 
-  protected ZooLock(ZooKeeper zookeeper, String path, UUID uuid) {
+  protected ZooLock(ZooKeeper zookeeper, ZooLockPath path, UUID uuid) {
```
Is line 177 correct, without IDE help it looks like path must be a String there.

Manno15 replied:
This line? You mention 177 but I see the method header in your snipit. 

https://github.com/apache/accumulo/blob/cac03f54971c2d2ebcfffd03ed593cf6a922b354/core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java#L177

EdColeman replied:
Yes - had issues getting GH to allow me to comment in the correct spot.

Manno15 replied:
Interesting. I see what you mean but I haven't noticed any issue with it and IntelliJ does not show an error either. I can add the `toString()` call for sanity reasons if you think that is best.

ctubbsii replied:
If the `toString()` is correct, then that line is correct. Java automatically calls `toString()` on non-null objects when doing string concatenation (and uses literal "null" if null).

EdColeman replied:
This is one reason that I prefer an explicit getPath() - if for any reason toString() was changed, it may be harder than it needs to be to find places where the behavior of toString value is being used, esp with an automatic conversion. But again, my opinion...

ctubbsii replied:
@EdColeman In general, I agree with you, on the need for explicit methods for serialization, rather than relying on `toString()`, but in this case, I don't think it makes sense, since these are merely more strongly typed strings, so `toString()` should always be implemented correctly. It's more intuitive to think of these as more strongly typed strings, than it is to consider them path object types that have getters to a contained path object type.

EdColeman commented on core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java:
```diff
@@ -42,6 +42,19 @@
+    }
+  }
+
   public ZooQueueLock(ZooReaderWriter zrw, String path, boolean ephemeral) {
```
Can this be converted to use FateLockPath?

Manno15 replied:
Yes, I overlooked this. Good catch, I will add this in my next commit.

Manno15 replied:
Made this change in my latest commit

EdColeman commented on core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java:
```diff
@@ -42,6 +42,19 @@
+      this.path = path;
+    }
+
+    @Override
```
Would getPath() improve readability vs. toString()?

Manno15 replied:
I did have a few different versions of this. One with getPath and I even had a method in `ZooLock` and `ZooQueueLock` to get the lock path at one point. They all seemed readable to me so I stuck with what seemed to be the simplest approach. If you think it adds good value to this then I can definitely add that back in. 

To be clear, are you suggesting renaming toString() to getPath() or something else?

EdColeman replied:
Probably depends when you ask me - on the first I saw toString() I wanted to confirm that it was just the path - getPath() would have been clearer to me and followed more of a getter / setter pattern (well getter since the class is immutable)  If there were additional convenience methods added, then it probably would make more sense.  Adding additional convenience functions if there are common code access patterns would be a stronger encapsulation of the (X)Lock concept as a object vs just a wrapper around a String.  But I think a lot of this is also personal preference - that's why I am trying to frame these questions to see if others had an opinion.

ctubbsii replied:
I think there are 3 identical threads of conversation about `toString()` vs. `getPath()`, so we don't keep responding to them separately, I'm just going to mark two of them as "Resolved", and if there's more to discuss, then it can be discussed on the one that is left "Unresolved".

EdColeman reviewed (commented):
From a quick look things look really good - just a few questions - and I'll probably make a second pass loading the PR into an IDE.

Manno15 commented:
Thanks for taking the time to review it!

ctubbsii commented on core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java:
```diff
@@ -60,6 +60,19 @@ public String toString() {
+    private final String path;
+
+    public ZooLockPath(String path) {
+      this.path = path;
```
```suggestion
      this.path = Objects.requireNonNull(path);
```

ctubbsii commented on core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooQueueLock.java:
```diff
@@ -39,10 +39,23 @@
+    private final String path;
+
+    public FateLockPath(String path) {
+      this.path = path;
```
```suggestion
      this.path = Objects.requireNonNull(path);
```

EdColeman commented:
I have no remaining issues once Christopher's suggestions are resolved.

ctubbsii commented:
I do have some follow-up suggestions I was working on that I thought would make sense as part of this PR. @Manno15 any problem if I push them directly to your branch for inclusion? We can revert if there's any objection.

Manno15 commented:
@ctubbsii Go ahead. I will review them in the morning.

Manno15 commented:
I like your additions. Some comments below, in order of the commit points. 

- The renaming was going to happen in #1967 which is why I didn't do it here. 
- I did have something akin to this in my original commit but wasn't quite sure if it was better or not. So I removed it in the next commit. I did it this way so I can drop the second one if I changed my mind about it. I  do personally like having the path function better though. 
- I am indifferent to the `var` and method name change. I agree the shorter name is probably better, especially with a comment on the method now to explain its intent.
- Good change, I thought the fields were already final so that was my mistake.
- Make sense if it wasn't used. Good change.
Everything else looks good.

ctubbsii commented:
> I like your additions. Some comments below, in order of the commit points.
> 
> * The renaming was going to happen in #1967 which is why I didn't do it here.

This PR can fix both #1966 and #1967. One PR to address two issues. I think `ServiceLock` is a good name for `ZooLock`, so if there aren't any objections to that name, I'll add that rename to this PR also, and then we can merge it and close both issues.

> * Good change, I thought the fields were already final so that was my mistake.

To be clear, my additions weren't corrective in any way. I didn't see any mistakes. My additions were supplemental only, building on what you did. :smiley_cat:

Manno15 commented:
> This PR can fix both #1966 and #1967. One PR to address two issues. I think ServiceLock is a good name for ZooLock, so if there aren't any objections to that name, I'll add that rename to this PR also, and then we can merge it and close both issues.

I am also okay with that name if no one else objects or has any other alternatives for the rename in #1967.

Manno15 commented:
@ctubbsii Everything looks good to me. Is there anything else that you think needs to be added? If not, I can merge this in.

Manno15 commented:
I also ran the IT suite and everything passes.

ctubbsii commented:
@Manno15 I merged it. Thanks!
